File: C++/Iterator/BookCollection/collection_iterator.cc
#include <stdexcept>
#include "collection_iterator.h"

ScienceCollectionIterator::ScienceCollectionIterator(const Book* books, int size) : _books(books), _size(size),
                                                                              _current_index(0) {}

void ScienceCollectionIterator::first() {
    _current_index  = 0;
}

void ScienceCollectionIterator::next() {
    _current_index++;
}

bool ScienceCollectionIterator::is_done() const {
    return _current_index >= _size;
}

const Book& ScienceCollectionIterator::current_item() const {
    if (is_done()) {
        throw std::runtime_error("Iterator out of bounds...");
    } else {
        return _books[_current_index];
    }
}

MusicCollectionIterator::MusicCollectionIterator(const std::vector<Book>& books) : _books(books),
                                                                                   _current_index(0) {}

void MusicCollectionIterator::first() {
    _current_index = 0;
}

void MusicCollectionIterator::next() {
    _current_index++;
}

bool MusicCollectionIterator::is_done() const {
    return _current_index >= _books.size();
}

const Book &MusicCollectionIterator::current_item() const {
    if (is_done()) {
        throw std::runtime_error("Iterator out of bounds...");
    } else {
        return _books[_current_index];
    }
}

File: Java/Iterator/BookCollection/src/com/ericgtkb/CollectionIterator.java
package com.ericgtkb;


import java.util.List;

// This is an implementation of the GoF iterator
// first moves the iterator to the first item
// next advances the iterator to the next element but doesn't return it
// currentItem returns current element
public interface CollectionIterator {
    void first();
    void next();
    boolean isDone();
    Book currentItem() throws Exception;
}

class ScienceCollectionIterator implements CollectionIterator {
    private Book[] books;
    private int currentIndex;

    public ScienceCollectionIterator(Book[] books) {
        this.books = books;
        currentIndex = 0;
    }

    @Override
    public void first() {
        currentIndex = 0;
    }

    @Override
    public void next() {
        currentIndex++;
    }

    @Override
    public boolean isDone() {
        return currentIndex >= books.length;
    }

    @Override
    public Book currentItem() throws Exception {
        if (isDone()) {
            throw new Exception("Iterator out of bounds...");
        } else {
            return books[currentIndex];
        }
    }
}

class MusicCollectionIterator implements CollectionIterator {
    private List<Book> bookList;
    private int currentIndex;

    public MusicCollectionIterator(List<Book> bookList) {
        this.bookList = bookList;
        currentIndex = 0;
    }

    @Override
    public void first() {
        currentIndex = 0;
    }

    @Override
    public void next() {
        currentIndex++;
    }

    @Override
    public boolean isDone() {
        return currentIndex >= bookList.size();
    }

    @Override
    public Book currentItem() throws Exception {
        if (isDone()) {
            throw new Exception("Iterator out of bounds...");
        } else {
            return bookList.get(currentIndex);
        }
    }
}

